### PR TITLE
netclient: init at 0.21.0

### PR DESCRIPTION
**Pull request by wexder**
###### Description of changes
Added new package netclient https://github.com/gravitl/netclient

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/netclient/default.nix [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+  vendorHash = "sha256-qchVMLmdvqpVpyjDy05W8yHTBcYCgpyuX5Ibg+yceXY=";
+
+  buildInputs = [
+    libglvnd
+    xorg.libX11
+    xorg.libXcursor
+    xorg.libXi
+    xorg.libXinerama
+    xorg.libXrandr
```
Comment (Gerg-L) [inline pkgs/applications/networking/netclient/default.nix:29]:
```suggestion
    libX11
    libXcursor
    libXi
    libXinerama
    libXrandr
```

**Review by Gerg-L — CHANGES_REQUESTED** [review]

**Review by Gerg-L — APPROVED** [review]

**Review by GiggleSquid — APPROVED** [review]

**Comment by GiggleSquid** [comment]
Result of `nixpkgs-review pr 244197` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>netclient</li>
  </ul>
</details>

File: pkgs/applications/networking/netclient/default.nix [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+, libXrandr
+}:
+
+buildGoModule rec {
+  pname = "netclient";
+  version = "0.20.3";
+
+  src = fetchFromGitHub {
+    owner = "gravitl";
+    repo = pname;
```
Comment (Lord-Valen) [inline pkgs/applications/networking/netclient/default.nix:18]:
```suggestion
    repo = "netclient";
```

**Comment by Lord-Valen** [comment]
From ofborg x86_64-darwin
```
trying https://github.com/gravitl/netclient/archive/v0.20.6.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  311k    0  311k    0     0   2401      0 --:--:--  0:02:13 --:--:--     0
curl: (92) HTTP/2 stream 1 was not closed cleanly: CANCEL (err 8)
error: cannot download source from any mirror
error: builder for '/nix/store/gsqsmbipncg3qm4s62ppiv4fnrlgv3s6-source.drv' failed with exit code 1;
       last 8 log lines:
       >
       > trying https://github.com/gravitl/netclient/archive/v0.20.6.tar.gz
       >   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
       >                                  Dload  Upload   Total   Spent    Left  Speed
       >   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
       > 100  311k    0  311k    0     0   2401      0 --:--:--  0:02:13 --:--:--     0
       > curl: (92) HTTP/2 stream 1 was not closed cleanly: CANCEL (err 8)
       > error: cannot download source from any mirror
       For full logs, run 'nix log /nix/store/gsqsmbipncg3qm4s62ppiv4fnrlgv3s6-source.drv'.
error: 1 dependencies of derivation '/nix/store/fxj7161d9x4yz6iw1s0bkn612j2gb9vv-netclient-0.20.6-go-modules.drv' failed to build
```

From ofborg aarch64-darwin
```
unpacking source archive /private/tmp/nix-build-source.drv-0/v0.20.6.tar.gz
gzip: stdin: unexpected end of file
tar: Unexpected EOF in archive
tar: Unexpected EOF in archive
tar: Error is not recoverable: exiting now
do not know how to unpack source archive /private/tmp/nix-build-source.drv-0/v0.20.6.tar.gz
error: builder for '/nix/store/is1qv86jnr5nr21nvxv3sz8mgm9mpd72-source.drv' failed with exit code 1;
       last 10 log lines:
       >                                  Dload  Upload   Total   Spent    Left  Speed
       >   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
       > 100 3392k    0 3392k    0     0   9442      0 --:--:--  0:06:07 --:--:--     0
       > unpacking source archive /private/tmp/nix-build-source.drv-0/v0.20.6.tar.gz
       >
       > gzip: stdin: unexpected end of file
       > tar: Unexpected EOF in archive
       > tar: Unexpected EOF in archive
       > tar: Error is not recoverable: exiting now
       > do not know how to unpack source archive /private/tmp/nix-build-source.drv-0/v0.20.6.tar.gz
       For full logs, run 'nix log /nix/store/is1qv86jnr5nr21nvxv3sz8mgm9mpd72-source.drv'.
error: 1 dependencies of derivation '/nix/store/jz7krwaxjmzd79ifiyi966rns7b85in3-netclient-0.20.6-go-modules.drv' failed to build
```

**Comment by wexder** [comment]
> From ofborg x86_64-darwin
> 
> ```
> trying https://github.com/gravitl/netclient/archive/v0.20.6.tar.gz
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> 100  311k    0  311k    0     0   2401      0 --:--:--  0:02:13 --:--:--     0
> curl: (92) HTTP/2 stream 1 was not closed cleanly: CANCEL (err 8)
> error: cannot download source from any mirror
> error: builder for '/nix/store/gsqsmbipncg3qm4s62ppiv4fnrlgv3s6-source.drv' failed with exit code 1;
>        last 8 log lines:
>        >
>        > trying https://github.com/gravitl/netclient/archive/v0.20.6.tar.gz
>        >   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>        >                                  Dload  Upload   Total   Spent    Left  Speed
>        >   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
>        > 100  311k    0  311k    0     0   2401      0 --:--:--  0:02:13 --:--:--     0
>        > curl: (92) HTTP/2 stream 1 was not closed cleanly: CANCEL (err 8)
>        > error: cannot download source from any mirror
>        For full logs, run 'nix log /nix/store/gsqsmbipncg3qm4s62ppiv4fnrlgv3s6-source.drv'.
> error: 1 dependencies of derivation '/nix/store/fxj7161d9x4yz6iw1s0bkn612j2gb9vv-netclient-0.20.6-go-modules.drv' failed to build
> ```
> 
> From ofborg aarch64-darwin
> 
> ```
> unpacking source archive /private/tmp/nix-build-source.drv-0/v0.20.6.tar.gz
> gzip: stdin: unexpected end of file
> tar: Unexpected EOF in archive
> tar: Unexpected EOF in archive
> tar: Error is not recoverable: exiting now
> do not know how to unpack source archive /private/tmp/nix-build-source.drv-0/v0.20.6.tar.gz
> error: builder for '/nix/store/is1qv86jnr5nr21nvxv3sz8mgm9mpd72-source.drv' failed with exit code 1;
>        last 10 log lines:
>        >                                  Dload  Upload   Total   Spent    Left  Speed
>        >   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
>        > 100 3392k    0 3392k    0     0   9442      0 --:--:--  0:06:07 --:--:--     0
>        > unpacking source archive /private/tmp/nix-build-source.drv-0/v0.20.6.tar.gz
>        >
>        > gzip: stdin: unexpected end of file
>        > tar: Unexpected EOF in archive
>        > tar: Unexpected EOF in archive
>        > tar: Error is not recoverable: exiting now
>        > do not know how to unpack source archive /private/tmp/nix-build-source.drv-0/v0.20.6.tar.gz
>        For full logs, run 'nix log /nix/store/is1qv86jnr5nr21nvxv3sz8mgm9mpd72-source.drv'.
> error: 1 dependencies of derivation '/nix/store/jz7krwaxjmzd79ifiyi966rns7b85in3-netclient-0.20.6-go-modules.drv' failed to build
> ```

Can you help me here ? If I understand it correctly, the source tar failed to download but only for darwin arch ? Why that could be ?

**Comment by Lord-Valen** [comment]
> @Lord-Valen Can you help me here ? If I understand it correctly, the source tar failed to download but only for darwin arch ? Why that could be ?

I'm not sure what caused that myself. Maybe it was a fluke?
@ofborg build netclient

**Comment by Lord-Valen** [comment]
Ah, it was for `passthru.tests`. The actual problem:
```
# golang.design/x/clipboard
vendor/golang.design/x/clipboard/clipboard_darwin.go:16:9: fatal error: 'Cocoa/Cocoa.h' file not found
#import <Cocoa/Cocoa.h>
        ^~~~~~~~~~~~~~~
1 error generated.
error: builder for '/nix/store/8qpa4gphzlvhbi1s6zcavml034zw4r1a-netclient-0.20.6.drv' failed with exit code 1;
       last 10 log lines:
       > Building subPackage ./config
       > Building subPackage ./daemon
       > Building subPackage ./firewall
       > Building subPackage ./functions
       > Building subPackage ./gui
       > # golang.design/x/clipboard
       > vendor/golang.design/x/clipboard/clipboard_darwin.go:16:9: fatal error: 'Cocoa/Cocoa.h' file not found
       > #import <Cocoa/Cocoa.h>
       >         ^~~~~~~~~~~~~~~
       > 1 error generated.
       For full logs, run 'nix log /nix/store/8qpa4gphzlvhbi1s6zcavml034zw4r1a-netclient-0.20.6.drv'.
   ```

**Comment by wexder** [comment]
https://github.com/ofborg build netclient

**Comment by wexder** [comment]
@ofborg build netclient

**Comment by wexder** [comment]
@ofborg build netclient

**Comment by wexder** [comment]
@ofborg build netclient

**Comment by wexder** [comment]
@ofborg build netclient

**Comment by wexder** [comment]
> Ah, it was for `passthru.tests`. The actual problem:
> 
> ```
> # golang.design/x/clipboard
> vendor/golang.design/x/clipboard/clipboard_darwin.go:16:9: fatal error: 'Cocoa/Cocoa.h' file not found
> #import <Cocoa/Cocoa.h>
>         ^~~~~~~~~~~~~~~
> 1 error generated.
> error: builder for '/nix/store/8qpa4gphzlvhbi1s6zcavml034zw4r1a-netclient-0.20.6.drv' failed with exit code 1;
>        last 10 log lines:
>        > Building subPackage ./config
>        > Building subPackage ./daemon
>        > Building subPackage ./firewall
>        > Building subPackage ./functions
>        > Building subPackage ./gui
>        > # golang.design/x/clipboard
>        > vendor/golang.design/x/clipboard/clipboard_darwin.go:16:9: fatal error: 'Cocoa/Cocoa.h' file not found
>        > #import <Cocoa/Cocoa.h>
>        >         ^~~~~~~~~~~~~~~
>        > 1 error generated.
>        For full logs, run 'nix log /nix/store/8qpa4gphzlvhbi1s6zcavml034zw4r1a-netclient-0.20.6.drv'.
> ```

@Lord-Valen  Can you please take a look if the buildInputs looks good. It makes sense to me to have it like this, but it might not be the correct way

File: pkgs/applications/networking/netclient/default.nix [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  src = fetchFromGitHub {
+    owner = "gravitl";
+    repo = "netclient";
+    rev = "v${version}";
+    hash = "sha256-68/BmVoAFaIg4vgjzhedSBqm6H9VDu3M7JemfPEcpjQ=";
+  };
+
+  vendorHash = "sha256-CsW4tW6+INw93A7uXtHeVnxRrE5unHXhm2SOmQkJwYA=";
+
+  buildInputs = [ ] ++ lib.optional stdenv.isDarwin darwin.apple_sdk.frameworks.Cocoa
```
Comment (Lord-Valen) [inline pkgs/applications/networking/netclient/default.nix:22]:
```suggestion
  buildInputs = lib.optional stdenv.isDarwin darwin.apple_sdk.frameworks.Cocoa
```

Comment (wexder) [reply pkgs/applications/networking/netclient/default.nix]: Thought this would be your suggestion. Anyway it's done

**Review by Lord-Valen — APPROVED** [review]

**Comment by Gerg-L** [comment]
LGTM

**Review by Lord-Valen — APPROVED** [review]

**Comment by purepani** [comment]
@wegank Do you think you can merge this?

**Comment by Gerg-L** [comment]
@wexder you have merge conflicts

**Comment by wexder** [comment]
@wegank thank you for resolving merge conflicts

**Review by wegank — APPROVED** [review]

**Pull request merged**